### ISSUE TITLE
Cannot get metadata on MQTT topic

### ISSUE THREAD

**Issue by rogersmj** [mikebrady/shairport-sync#1380]
Currently running latest development branch with AirPlay 2. Configured with:

`./configure --sysconfdir=/etc --with-alsa --with-soxr --with-avahi --with-ssl=opensll --with-systemd --with-airplay-2 --with-mqtt-client --with-metadata`

Running an MQTT Explorer tool on my Mac (connected to the MQTT broker using the same credentials as in my shairport-sync configuration file, I don't see anything coming through on the topic I've setup for shairport-sync. Here's the relevant parts of the conf file:

```
metadata =
{
        enabled = "yes"; // set this to yes to get Shairport Sync to solicit metadata from the source and to pass it on via a pipe
        include_cover_art = "yes"; // set to "yes" to get Shairport Sync to solicit cover art from the source and pass it via the pipe. You must also set "enabled" to "yes".
        cover_art_cache_directory = "/tmp/shairport-sync/.cache/coverart"; // artwork will be  stored in this directory if the dbus or MPRIS interfaces are enabled or if the MQTT client is i$
        pipe_name = "/tmp/shairport-sync-metadata";
        pipe_timeout = 5000; // wait for this number of milliseconds for a blocked pipe to unblock before giving up
//      socket_address = "226.0.0.1"; // if set to a host name or IP address, UDP packets containing metadata will be sent to this address. May be a multicast address. "socket-port" must be $
//      socket_port = 5555; // if socket_address is set, the port to send UDP packets to
//      socket_msglength = 65000; // the maximum packet size for any UDP metadata. This will be clipped to be between 500 or 65000. The default is 500.
};

// How to enable the MQTT-metadata/remote-service
// For this section to be operative, Shairport Sync must be built with the following configuration flag:
// --with-mqtt-client
mqtt =
{
        enabled = "yes"; // set this to yes to enable the mqtt-metadata-service
        hostname = "192.168.2.2"; // Hostname of the MQTT Broker
        port = 1883; // Port on the MQTT Broker to connect to
        username = "REDACTED"; //set this to a string to your username in order to enable username authentication
        password = "REDACTED"; //set this to a string you your password in order to enable username & password authentication
        topic = "shairport"; //MQTT topic where this instance of shairport-sync should publish. If not set, the general.name value is used.
//      publish_raw = "no"; //whether to publish all available metadata under the codes given in the 'metadata' docs.
        publish_parsed = "yes"; //whether to publish a small (but useful) subset of metadata under human-understandable topics
//      publish_cover = "no"; //whether to publish the cover over mqtt in binary form. This may lead to a bit of load on the broker
        enable_remote = "yes"; //whether to remote control via MQTT. RC is available under `topic`/remote.

};
```

At some point in the past, this worked, and I'm not sure when it stopped working. When I run shairport-sync from the command line with debug info, I don't see anything immediately obvious in the output:

```
         0.002082812 "shairport.c:1820" Features: 0x1c340405fca00.
         0.000211615 "shairport.c:1849" Started in Airplay 2 mode on device "B8:27:EB:21:2C:78"!
         0.000072812 "shairport.c:1891" software version: "4.1-dev-110-g6f7e8947-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-sysconfdir:/etc"
         0.000024219 "shairport.c:1897" log verbosity is 1.
         0.000057708 "audio_alsa.c:1018" alsa: alsa_maximum_stall_time of 0.200000 sec.
         0.000065469 "audio_alsa.c:1305" alsa: disable_standby_mode is "never".
         0.000022500 "audio_alsa.c:1307" alsa: disable_standby_mode_silence_threshold is 0.040000 seconds.
         0.000024063 "audio_alsa.c:1309" alsa: disable_standby_mode_silence_scan_interval is 0.004000 seconds.
         0.000034583 "audio_alsa.c:1348" alsa: output device name is "hw:Headphones".
         0.000325521 "shairport.c:1926" libsodium initialised.
         0.000050729 "shairport.c:1976" disable resend requests is off.
         0.000023125 "shairport.c:1980" diagnostic_drop_packet_fraction is 0.000000. A value of 0.0 means no packets will be dropped deliberately.
         0.000025417 "shairport.c:1981" statistics_requester status is 1.
         0.000022708 "shairport.c:1986" rtsp listening port is 7000.
         0.000021146 "shairport.c:1987" udp base port is 6001.
         0.000021354 "shairport.c:1988" udp port range is 10.
         0.000021146 "shairport.c:1989" player name is "_House Audio".
         0.000019739 "shairport.c:1990" backend is "alsa".
         0.000018855 "shairport.c:1991" run_this_before_play_begins action is "(null)".
         0.000021406 "shairport.c:1992" run_this_after_play_ends action is "(null)".
         0.000019739 "shairport.c:1993" wait-cmd status is 0.
         0.000018907 "shairport.c:1994" run_this_before_play_begins may return output is 0.
         0.000019791 "shairport.c:1995" run_this_if_an_unfixable_error_is_detected action is "(null)".
         0.000022084 "shairport.c:1996" run_this_before_entering_active_state action is  "(null)".
         0.000019948 "shairport.c:1997" run_this_after_exiting_active_state action is  "(null)".
         0.000021406 "shairport.c:1998" active_state_timeout is  10.000000 seconds.
         0.000026041 "shairport.c:1999" mdns backend "(null)".
         0.000021042 "shairport.c:2004" interpolation setting is "auto".
         0.000021354 "shairport.c:2005" interpolation soxr_delay_threshold is 30.
         0.000019636 "shairport.c:2006" resync time is 0.050000 seconds.
         0.000022343 "shairport.c:2007" allow a session to be interrupted: 0.
         0.000021355 "shairport.c:2008" busy timeout time is 0.
         0.000021145 "shairport.c:2009" drift tolerance is 0.001995 seconds.
         0.000022448 "shairport.c:2010" password is "(null)".
         0.000020261 "shairport.c:2011" ignore_volume_control is 0.
         0.000021041 "shairport.c:2015" volume_max_db is not set
         0.000018594 "shairport.c:2017" volume range in dB (zero means use the range specified by the mixer): 60.
         0.000021302 "shairport.c:2021" volume_range_combined_hardware_priority (1 means hardware mixer attenuation is used first) is 0.
         0.000022813 "shairport.c:2023" playback_mode is 0 (0-stereo, 1-mono, 1-reverse_stereo, 2-both_left, 3-both_right).
         0.000022708 "shairport.c:2024" disable_synchronization is 0.
         0.000021042 "shairport.c:2025" use_mmap_if_available is 1.
         0.000021198 "shairport.c:2027" output_format automatic selection is enabled.
         0.000021614 "shairport.c:2031" output_rate automatic selection is enabled.
         0.000021250 "shairport.c:2035" audio backend desired buffer length is 0.200000 seconds.
         0.000023021 "shairport.c:2037" audio_backend_buffer_interpolation_threshold_in_seconds is 0.120000 seconds.
         0.000023698 "shairport.c:2038" audio backend latency offset is 0.000000 seconds.
         0.000022865 "shairport.c:2040" audio backend silence lead-in time is "auto".
         0.000021093 "shairport.c:2044" zeroconf regtype is "_raop._tcp".
         0.000021459 "shairport.c:2045" decoders_supported field is 1.
         0.000020937 "shairport.c:2046" use_apple_decoder is 0.
         0.000021406 "shairport.c:2047" alsa_use_hardware_mute is 0.
         0.000021303 "shairport.c:2051" no special mdns service interface was requested.
         0.000055520 "shairport.c:2055" configuration file name "/etc/shairport-sync.conf" resolves to "/etc/shairport-sync.conf".
         0.000026355 "shairport.c:2061" metadata enabled is 1.
         0.000022395 "shairport.c:2062" metadata pipename is "/tmp/shairport-sync-metadata".
         0.000022084 "shairport.c:2064" metadata socket address is "(null)" port 0.
         0.000022552 "shairport.c:2065" metadata socket packet size is "500".
         0.000021094 "shairport.c:2066" get-coverart is 1.
         0.000020781 "shairport.c:2084" loudness is 0.
         0.000020521 "shairport.c:2085" loudness reference level is -20.000000
         0.000222760 "rtsp.c:3833" metadata pipe name is "/tmp/shairport-sync-metadata".
         1.503956302 "shairport.c:224" "soxr" interpolation has been chosen.
         0.779780312 "rtsp.c:5195" Connection 1: new connection from fd8e:13c1:357:1:8df:6eef:8740:db3e:50415 to self at fd8e:13c1:357:1:2a8c:102a:c2d2:b463:7000.
        14.236082130 "rtsp.c:5195" Connection 2: new connection from fd8e:13c1:357:1:8df:6eef:8740:db3e:50422 to self at fd8e:13c1:357:1:2a8c:102a:c2d2:b463:7000.
         0.397374843 "ptp-utilities.c:151" Send control message to NQPTP: "T fd8e:13c1:357:1:8df:6eef:8740:db3e fe80::86f:2848:1a26:8b13 192.168.1.55"
         4.435712811 "rtsp.c:3040" SETUP on Connection 2. Buffered Audio Stream Detected.
         0.007951979 "audio_alsa.c:903" Lowest dB value is a mute
         0.001258802 "audio_alsa.c:529" alsa: output format chosen is "S16".
         0.000128438 "audio_alsa.c:571" alsa: output speed chosen is 44100.
         0.019748594 "rtp.c:1432" Connection 2: NQPTP clock interface mismatch.
         0.000134687 "rtp.c:1434" *warning: This version of Shairport Sync is not compatible with the installed version of NQPTP. Please update.
         0.000426563 "rtp.c:1432" Connection 2: NQPTP clock interface mismatch.
         0.000406562 "rtp.c:1434" *warning: This version of Shairport Sync is not compatible with the installed version of NQPTP. Please update.
```

**Comment by mikebrady** [comment]
Hi there. You need to update Shairport Sync _and_ NQPTP to the latest versions. AFAIK MQTT is working okay.

**Comment by rogersmj** [comment]
Hi @mikebrady -- after posting this I did update NQPTP and I'm still having the same problem and same warnings from shairport-sync about the NQPTP version:

```
 $ nqptp -V
Version: 1.1-dev-99-g4d9d431. Shared Memory Interface Version: 6.
$ shairport-sync -V
4.1-dev-110-g6f7e8947-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-sysconfdir:/etc
```

**Comment by rogersmj** [comment]
Well apparently I never unloaded the nqptp service and restarted it properly after the update. Now that I've done that, the warning about the incompatible NQPTP version has gone away, but I'm still not getting anything on my MQTT broker. Here's the latest runtime log from shairport-sync:

```
$ shairport-sync -vu
         0.002051354 "shairport.c:1820" Features: 0x1c340405fca00.
         0.000212812 "shairport.c:1849" Started in Airplay 2 mode on device "B8:27:EB:21:2C:78"!
         0.000071615 "shairport.c:1891" software version: "4.1-dev-110-g6f7e8947-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-sysconfdir:/etc"
         0.000023698 "shairport.c:1897" log verbosity is 1.
         0.000095885 "audio_alsa.c:1018" alsa: alsa_maximum_stall_time of 0.200000 sec.
         0.000100104 "audio_alsa.c:1305" alsa: disable_standby_mode is "never".
         0.000053855 "audio_alsa.c:1307" alsa: disable_standby_mode_silence_threshold is 0.040000 seconds.
         0.000054791 "audio_alsa.c:1309" alsa: disable_standby_mode_silence_scan_interval is 0.004000 seconds.
         0.000092917 "audio_alsa.c:1348" alsa: output device name is "hw:Headphones".
         0.000352083 "shairport.c:1926" libsodium initialised.
         0.000084063 "shairport.c:1976" disable resend requests is off.
         0.000054271 "shairport.c:1980" diagnostic_drop_packet_fraction is 0.000000. A value of 0.0 means no packets will be dropped deliberately.
         0.000059583 "shairport.c:1981" statistics_requester status is 0.
         0.000054062 "shairport.c:1986" rtsp listening port is 7000.
         0.000052136 "shairport.c:1987" udp base port is 6001.
         0.000051562 "shairport.c:1988" udp port range is 10.
         0.000043177 "shairport.c:1989" player name is "_House Audio".
         0.000020834 "shairport.c:1990" backend is "alsa".
         0.000019219 "shairport.c:1991" run_this_before_play_begins action is "(null)".
         0.000047656 "shairport.c:1992" run_this_after_play_ends action is "(null)".
         0.000051250 "shairport.c:1993" wait-cmd status is 0.
         0.000050625 "shairport.c:1994" run_this_before_play_begins may return output is 0.
         0.000052031 "shairport.c:1995" run_this_if_an_unfixable_error_is_detected action is "(null)".
         0.000055000 "shairport.c:1996" run_this_before_entering_active_state action is  "(null)".
         0.000052448 "shairport.c:1997" run_this_after_exiting_active_state action is  "(null)".
         0.000051510 "shairport.c:1998" active_state_timeout is  10.000000 seconds.
         0.000060677 "shairport.c:1999" mdns backend "(null)".
         0.000050990 "shairport.c:2004" interpolation setting is "auto".
         0.000050677 "shairport.c:2005" interpolation soxr_delay_threshold is 30.
         0.000055729 "shairport.c:2006" resync time is 0.050000 seconds.
         0.000057188 "shairport.c:2007" allow a session to be interrupted: 0.
         0.000055469 "shairport.c:2008" busy timeout time is 0.
         0.000054687 "shairport.c:2009" drift tolerance is 0.001995 seconds.
         0.000058698 "shairport.c:2010" password is "(null)".
         0.000051198 "shairport.c:2011" ignore_volume_control is 0.
         0.000051614 "shairport.c:2015" volume_max_db is not set
         0.000050573 "shairport.c:2017" volume range in dB (zero means use the range specified by the mixer): 60.
         0.000052344 "shairport.c:2021" volume_range_combined_hardware_priority (1 means hardware mixer attenuation is used first) is 0.
         0.000053438 "shairport.c:2023" playback_mode is 0 (0-stereo, 1-mono, 1-reverse_stereo, 2-both_left, 3-both_right).
         0.000052239 "shairport.c:2024" disable_synchronization is 0.
         0.000051615 "shairport.c:2025" use_mmap_if_available is 1.
         0.000054896 "shairport.c:2027" output_format automatic selection is enabled.
         0.000051562 "shairport.c:2031" output_rate automatic selection is enabled.
         0.000051458 "shairport.c:2035" audio backend desired buffer length is 0.200000 seconds.
         0.000052553 "shairport.c:2037" audio_backend_buffer_interpolation_threshold_in_seconds is 0.120000 seconds.
         0.000058437 "shairport.c:2038" audio backend latency offset is 0.000000 seconds.
         0.000056927 "shairport.c:2040" audio backend silence lead-in time is "auto".
         0.000055469 "shairport.c:2044" zeroconf regtype is "_raop._tcp".
         0.000055000 "shairport.c:2045" decoders_supported field is 1.
         0.000054948 "shairport.c:2046" use_apple_decoder is 0.
         0.000107239 "shairport.c:2047" alsa_use_hardware_mute is 0.
         0.000053386 "shairport.c:2051" no special mdns service interface was requested.
         0.000091041 "shairport.c:2055" configuration file name "/etc/shairport-sync.conf" resolves to "/etc/shairport-sync.conf".
         0.000058959 "shairport.c:2061" metadata enabled is 1.
         0.000051458 "shairport.c:2062" metadata pipename is "/tmp/shairport-sync-metadata".
         0.000052604 "shairport.c:2064" metadata socket address is "(null)" port 0.
         0.000052500 "shairport.c:2065" metadata socket packet size is "500".
         0.000056146 "shairport.c:2066" get-coverart is 1.
         0.000054844 "shairport.c:2084" loudness is 0.
         0.000051719 "shairport.c:2085" loudness reference level is -20.000000
         0.000267656 "rtsp.c:3833" metadata pipe name is "/tmp/shairport-sync-metadata".
         1.501558072 "shairport.c:224" "soxr" interpolation has been chosen.
         9.160443382 "rtsp.c:5195" Connection 1: new connection from fd8e:13c1:357:1:8df:6eef:8740:db3e:50588 to self at fd8e:13c1:357:1:2a8c:102a:c2d2:b463:7000.
         0.534805729 "ptp-utilities.c:151" Send control message to NQPTP: "T fd8e:13c1:357:1:8df:6eef:8740:db3e fe80::86f:2848:1a26:8b13 192.168.1.55"
         0.383149271 "rtsp.c:3040" SETUP on Connection 1. Buffered Audio Stream Detected.
         0.006658333 "audio_alsa.c:903" Lowest dB value is a mute
         0.001275833 "audio_alsa.c:529" alsa: output format chosen is "S16".
         0.000131355 "audio_alsa.c:571" alsa: output speed chosen is 44100.
         1.534170780 "rtp.c:1325" Connection 1: Clock 6c4a8569e7a70008 is now the new anchor clock and master clock. History: 2000.726301 milliseconds.
         0.000175990 "rtp.c:1408" Connection 1: NQPTP new master clock 6c4a8569e7a70008.
```

**Comment by mikebrady** [comment]
Good stuff.

So, by coincidence, there is a just-closed issue on MQTT at #1375. AFAICT it's working pretty okay. IS there anything there that would help?

**Comment by mikebrady** [comment]
By the way, that most recent build of Shairport Sync -- the one you just got working -- doesn't seem to include MQTT support:
```
$ shairport-sync -vu
         0.002051354 "shairport.c:1820" Features: 0x1c340405fca00.
         0.000212812 "shairport.c:1849" Started in Airplay 2 mode on device "B8:27:EB:21:2C:78"!
         0.000071615 "shairport.c:1891" software version: "4.1-dev-110-g6f7e8947-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-sysconfdir:/etc"
         0.000023698 "shairport.c:1897" log verbosity is 1.
...
```

**Comment by rogersmj** [comment]
Weird...I definitely included the correct build flags to enable MQTT. Based on that though, I stopped the service and started over, beginning with a `make clean`, then went through the configuration/build/install process again. Now MQTT works. I'm guessing when I upgraded from the older version I didn't clear out some compiled files and make skipped over those or something? I dunno but with a `make clean` MQTT works now. Closing.
